kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/IndexedIntInterval.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A non-empty integer interval. The index for each value is the 0-based position of the
+ * value in the interval from smallest to largest.
+ * @param startInclusive the inclusive start integer
+ * @param endExclusive the exclusive end integer
+ */
+record IndexedIntInterval(int startInclusive, int endExclusive) implements IndexedIntSet {
+
+    private static final String START_INCLUSIVE = "[";
+    private static final String START_EXCLUSIVE = "(";
+    private static final String END_INCLUSIVE = "]";
+    private static final String END_EXCLUSIVE = ")";
+    private static final String INTEGER_GROUP = "(\\d*)";
+    private static final String END_INCLUSIVITY_GROUP = "([" + END_EXCLUSIVE + Pattern.quote(END_INCLUSIVE) + "])";
+    private static final String START_INCLUSIVITY_GROUP = "([" + Pattern.quote(START_INCLUSIVE) + START_EXCLUSIVE + "])";
+    static Pattern INTERVAL_PATTERN = Pattern.compile("^" + START_INCLUSIVITY_GROUP + INTEGER_GROUP + "," + INTEGER_GROUP + END_INCLUSIVITY_GROUP + "$");
+
+    IndexedIntInterval {
+        if (endExclusive <= startInclusive) {
+            throw new IllegalArgumentException("endExclusive " + endExclusive + " less than or equal to startInclusive " + startInclusive);
+        }
+    }
+
+    @Override
+    public int indexOf(int integer) {
+        return contains(integer) ? integer - startInclusive : -1;
+    }
+
+    @Override
+    public boolean contains(int integer) {
+        return integer >= startInclusive && integer < endExclusive;
+    }
+
+    @Override
+    public int size() {
+        return endExclusive - startInclusive;
+    }
+
+    enum Inclusivity {
+        INCLUSIVE {
+            @Override
+            public int toStartInclusive(int value) {
+                return value;
+            }
+
+            @Override
+            public int toEndExclusive(int value) {
+                return value + 1;
+            }
+        },
+        EXCLUSIVE {
+            @Override
+            public int toStartInclusive(int value) {
+                return value + 1;
+            }
+
+            @Override
+            public int toEndExclusive(int value) {
+                return value;
+            }
+        };
+
+        public abstract int toStartInclusive(int value);
+
+        public abstract int toEndExclusive(int value);
+    }
+
+    /**
+     * Parse an integer range spec like [0,1). Square brackets are inclusive, round are exclusive. e.g.
+     * <ul>
+     *     <li>[1,3] -> (1,2,3)</li>
```

Comment:
I would use a different notation for the RHS (`{1, 2, 3}`) to make it clear you are not describing a range.